_episodes/l1-02-branching-merging.md

```diff
@@ -0,0 +1,578 @@
+---
+title: "Branching and Merging"
+teaching: 40
+exercises: 20
+questions:
+- How can I or my team work on multiple features in parallel?
+- How can changes from parallel tracks of work be combined?
+objectives:
+- Explain what git branches are and when they should be used
+- Use a branch to develop a new feature and incorporate it into your code
+- Identify the branches in a project and which branch is currently in use
+- Describe a scalable workflow for development with git
+keypoints:
+- Git allows non-linear commit histories called branches
+- A branch can be thought of as a label that applies to set of commits
+- Branches can and should be used to carry out development of new features
+- Branches in a project can be listed with git branch and created with git branch branch_name
+- The HEAD refers to the current position of the project in its commit history
+- The current branch can be changed using git checkout branch_name
+- Once a branch is complete the changes made can be integrated into the project using git merge branch_name
+- Merging creates a new commit in the target branch incorporating all of the changes made in a branch
+- Conflicts arise when two branches contain incompatible sets of changes and must be resolved before a merge can complete
+- Identify the details of merge conflicts using git diff and/or git status
+- A merge conflict can be resolved by manual editing followed by git add [conflicted file]... and git commit -m "commit_message"
+---
+
+## Motivation for branches
+
+For simple projects, working with a single branch where you keep adding commits is good
+enough. But chances are that you will want to unleash all the power of `git` at some
+point and start using branches.
+
+In a linear history, we have something like:
+
+![Linear]({{ site.baseurl }}/fig/branch1.png "Linear git
+repository"){:class="img-responsive"}
+
+* Commits are depicted here as little boxes with abbreviated hashes.
+* Here the branch `main` points to a commit.
+* "HEAD" is the current position (remember the recording head of tape
+  recorders?).
+* When we talk about branches, we often mean all parent commits, not only the
+  commit pointed to.
+
+### Now we want to do this:
+
+<div style="border:2px solid #000000;"> <img src="{{ site.baseurl
+  }}/fig/octopus.jpeg" width="60%"> <br> Source: <a
+  href="https://twitter.com/jay_gee/status/703360688618536960">https://twitter.com/jay_gee/status/703360688618536960</a>
+  </div>
+
+Software development is often not linear:
+
+* We typically need at least one version of the code to "work" (to compile, to
+  give expected results, ...).
+* At the same time we work on new features, often several features concurrently.
+  Often they are unfinished.
+* We need to be able to separate different lines of work really well.
```

Comment:
```suggestion
* We need to be able to easily separate out work on different features.
```

Just a suggestion: on first reading I thought the "lines of work" maybe referred to lines of code or something.